text/3778-proc-macro-in-same-package-as-app.md

```diff
@@ -0,0 +1,95 @@
+
+This proposal aims smooth out the user experience when it comes to creating new proc macro, and achieve a similar effect to the F2 operation. It is important to emphasise that proc macros can dramatically simplify code, especially derive macros, but they a lot of the times aren't used because of all the extra hoops one has to get through. This would make proc macros (more of) "yet another feature", rather than a daunting one.
+
+An objection to this one might raise is "How much harder is typing in `cargo new` than `mkdir proc-macro`?" But we should consider if we would still use as much integration tests if the `tests` directory if it is required to be in a seperate package. The answer is most likely less. This is because (1) having a new package requires ceremony, like putting in a new dependency in cargo.toml, and (2) requires adding to the project structure. A *tiny* bit in lowering the interaction cost, even from 2 steps to 1, can greatly improve the user experience. 
+
+In summary (TL;DR), the effort one needs to put in to use a feature is extremely important. Proc macros currently has a higher ceiling, needing one to create a whole new package in order to use it, and lowering the ceiling, even just a little bit, could massively improve user experience. This proposal can lower it.
+
+# Explanation
+[explanation]: #explanation
+
+Currently, we create a new proc macro as so:
+1. Create a new package
+2. In its cargo.toml, specify that it is a proc macro package
+3. In the main project, add the package as a dependency
+4. Implement the proc macro in the new package
+
+After this change, we create a new proc macro like this:
+1. Create a new directory called `proc-macro` alongside your `src` directory
+2. Implement the proc macro in a new file in `proc-macro`.
+
+To use the proc macro, simply import it via `crate::proc_macro`.
+```rust
+use crate::proc_macro::my_file::my_macro;
+```
+Or, if the file happens to be `mod.rs`, you can access it directly after the `proc_macro` bit.
+
+## Proc Macro Libraries
+Libraries like `syn`, `quote`, and `proc-macro2`, would be included under `[dev-dependecies]` in the cargo.toml. (Perhaps we should put it in build dependencies? or a new dependency section for proc macros.)
+
+## How it would work in the implementation
+Cargo would have to compile the `proc-macro` directory first, as a proc macro type (of course). Then, in compiling the main code, `crate::proc_macro::file_name::my_macro` would resolve the module to the file `proc-macro/file_name.rs`. Alternatively, if the user uses `mod.rs`, it would be resolved from `crate::proc_macro::my_macro`. This would finally be passed into rustc.
```

Comment:
From https://github.com/rust-lang/rfcs/pull/3826#pullrequestreview-2881560746

> Maybe the RFC should also specify a few more things such as in what order it should be compiled compared to the build.rs. It could either be compiled:
> 
> * in parallel to the build.rs, to optimize for compile time.
> 
> * or before the build.rs, so that build.rs could use the macros.
> 
> * or after running the build.rs script, so that metadata and things generated in OUT_DIR would be available. (my preference)


(split in a thread to make it not lost)

Reply:
> For some code, the main crate depends on the proc-macro so it can both export it and use it internally, e.g.: [fayalite](https://git.libre-chip.org/libre-chip/fayalite)

IMO it should be configurable then. Having this dependency by default when not necessary and reducing pipelining (and hence increasing compile times) would be unfortunate.

> It has to be built before the main crate because it is built before lib.

What are "main crate" and "lib" for you? They are the same for me. And even if they were different, why does the proc macro need to be built before the lib?

Reply:
Do we have any future possibilities that rustc allows building, for example, most of serde except the parts that rely on `serde_derive`, then wait for cargo to link it with serde_derive in a final step, producing serde as an intact crate? If rustc cannot support this, what we get from this RFC will only be a superset of the problems that serde already has.

Reply:
I assume you are referring to the workaround of people depending on `serde_derive` directly and this solution not allowing that?

imo the correct way of solving that is to pull out a `serde_core` which is happening in serde-rs/serde#2608 

This is discussed more in the thread https://github.com/rust-lang/rfcs/pull/3826#discussion_r2129526621